File: mariadb-api/src/Services/PortalSlugResolver.php
<?php

declare(strict_types=1);

namespace Enkl\Api\Services;

use PDO;

/**
 * Ported from php-api/src/Services/PortalSlugResolver.php (itself ported from
 * Services/PortalSlugResolver.cs). Mirrors the project-key derive-then-uniquify shape for Portal
 * Slug instead of Project Key — a human-readable, hashbang-routable (#!/portal/<slug>) identifier,
 * unique per Organisation. No dialect divergence from the Postgres tier anywhere in this file.
 */
final class PortalSlugResolver
{
    public static function deriveSlug(?string $requestedSlug, string $name): string
    {
        $fromRequested = self::slugify($requestedSlug ?? '');
        if ($fromRequested !== '') {
            return self::truncate($fromRequested);
        }
        $fromName = self::slugify($name);
        return self::truncate($fromName !== '' ? $fromName : 'portal');
    }

    public static function resolveUniqueSlug(PDO $db, string $baseSlug, string $organisationId, ?string $excludePortalId = null): string
    {
        $sql = 'SELECT 1 FROM "Portals" WHERE "Slug" = :slug AND "OrganisationId" = :orgId'
            . ($excludePortalId !== null ? ' AND "Id" != :excludeId' : '');
        $candidate = $baseSlug;
        $suffix = 1;
        while (true) {
            $params = ['slug' => $candidate, 'orgId' => $organisationId];
            if ($excludePortalId !== null) {
                $params['excludeId'] = $excludePortalId;
            }
            $stmt = $db->prepare($sql);
            $stmt->execute($params);
            if ($stmt->fetch() === false) {
                return $candidate;
            }
            $suffix++;
            $candidate = "{$baseSlug}-{$suffix}";
        }
    }

    private static function slugify(string $value): string
    {
        $lowered = mb_strtolower(trim($value));
        $withDashes = preg_replace('/[^a-z0-9]+/', '-', $lowered) ?? '';
        return trim($withDashes, '-');
    }

    private static function truncate(string $slug): string
    {
        return mb_strlen($slug) > 80 ? rtrim(mb_substr($slug, 0, 80), '-') : $slug;
    }
}

File: api/Enkl.Api/Services/PortalSlugResolver.cs
using Enkl.Api.Data;
using Microsoft.EntityFrameworkCore;

namespace Enkl.Api.Services;

/// <summary>
/// Mirrors ProjectKeyResolver's derive-then-uniquify shape, for Portal.Slug instead of Project.Key —
/// a human-readable, hashbang-routable (#!/portal/&lt;slug&gt;) identifier, unique per Organisation.
/// </summary>
public static class PortalSlugResolver
{
    public static string DeriveSlug(string? requestedSlug, string name)
    {
        var fromRequested = Slugify(requestedSlug ?? "");
        if (fromRequested.Length > 0) return Truncate(fromRequested);

        var fromName = Slugify(name);
        return Truncate(fromName.Length > 0 ? fromName : "portal");
    }

    public static async Task<string> ResolveUniqueSlugAsync(AppDbContext db, string baseSlug, Guid organisationId, Guid? excludePortalId = null)
    {
        var candidate = baseSlug;
        var suffix = 1;
        while (await db.Portals.AnyAsync(p => p.Slug == candidate && p.OrganisationId == organisationId && p.Id != excludePortalId))
        {
            candidate = $"{baseSlug}-{++suffix}";
        }
        return candidate;
    }

    private static string Slugify(string value)
    {
        var lowered = value.Trim().ToLowerInvariant();
        var chars = new char[lowered.Length];
        var len = 0;
        var lastWasDash = false;
        foreach (var c in lowered)
        {
            if (char.IsLetterOrDigit(c))
            {
                chars[len++] = c;
                lastWasDash = false;
            }
            else if (!lastWasDash && len > 0)
            {
                chars[len++] = '-';
                lastWasDash = true;
            }
        }
        if (len > 0 && chars[len - 1] == '-') len--;
        return new string(chars, 0, len);
    }

    private static string Truncate(string slug) => slug.Length > 80 ? slug[..80].TrimEnd('-') : slug;
}
